### PR TITLE
Repositories Refresh Interval

### LINKED ISSUE #16852: Units mismatch for ClusterRepo refreshInterval
**Setup**
- Rancher version: 2.14 (and maybe older)

**Describe the bug**
Editing a `ClusterRepo` object (from the `catalog.cattle.io/v1` API group) through the UI shows "hours" units in the `Refresh Interval` input box, which then gets directly used as is in the `PUT` request to update the resource. However, the correct unit for this field is Seconds.

**To Reproduce**
Go to any Cluster's explorer -> Apps -> Repositories -> click any, e.g. "Rancher" (`rancher-charts`) -> "⋮" -> Edit Config -> (open Browser Dev Tools to inspect the network call) -> input "2" in the "Refresh interval" field and click Save.

**Result**
Inspect the request payload, and observe how ".spec.refreshInterval" is set to a value of "2" (not the equivalent in seconds for 2 hours).

**Expected Result**
Either the UI is changed to reflect the correct units, or it does the appropriate conversion to seconds.

**Screenshots**
Image
Image

**Additional context**
 - [Rancher code where the integer is interpreted as seconds](https://github.com/rancher/rancher/blob/main/pkg/controllers/dashboard/helm/repo.go#L118)

### PR DESCRIPTION
**Pull request by momesgin**
### Summary
Fixes #16852

The ClusterRepo edit form displayed "hours" as the refresh interval unit but sent the raw number as seconds in the API payload. This PR replaces the `UnitInput` with a `LabeledInput` + `LabeledSelect` combo so users can choose between seconds, minutes, hours, and days. A checkbox allows disabling refresh entirely (sets value to `-1`). The detail page now shows a human-readable duration.

### Occurred changes and/or fixed issues
- Replaced `UnitInput` with `LabeledInput` (number) + `LabeledSelect` (unit picker) + `Checkbox` (enable/disable)
- Added `secondsToLargestUnit` and `formatDuration` helpers to `shell/utils/time.ts`
- Added `refreshIntervalDisplay`, `defaultRefreshInterval`, and `defaultRefreshIntervalHours` getters to the ClusterRepo model
- Added "Refresh Interval" to the detail page via the `details` getter
- Added input validation rejecting negative values
- Updated translations and Cypress page objects/tests

### Technical notes summary
- The refresh interval is stored in seconds on `spec.refreshInterval`. The UI decomposes it into a display value and unit multiplier (1/60/3600/86400) on load via `secondsToLargestUnit`, and recomposes on save via `displayValue * unit`.
- Setting `spec.refreshInterval` to `-1` disables periodic updates. Clearing the input deletes the field from the spec, letting the backend use its default.
- Existing stored values are parsed into the largest fitting unit when editing.

### Areas or cases that should be tested
- Create/edit each repo type (Helm, Git, OCI) and verify the refresh interval value in the PUT/POST payload is in seconds
- Toggle the "Enabled" checkbox and verify `-1` is sent when disabled
- Change units and verify the correct multiplier is applied
- Edit an existing repo with a stored refresh interval and verify it decomposes correctly
- Verify the detail page shows a human-readable duration
- Verify negative input shows a validation error

### Areas which could experience regressions
- ClusterRepo create/edit form for all repo types (Helm, Git, OCI, SUSE App Collection)
- ClusterRepo detail page (new "Refresh Interval" field in details)

### Screenshot/Video

https://github.com/user-attachments/assets/31b21c98-55db-4014-9bcd-13b79e2efd45

UPDATE: the latest UI for the refresh interval section should look like this:
image

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone 
- [x] The PR template has been filled out
- [x] The PR has been self reviewed 
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes
- [x] The PR has been reviewed in terms of Accessibility
- [x] The PR has considered, and if applicable tested with, the three Global Roles `Admin`, `Standard User` and `User Base`

### CODE REVIEW

File: shell/assets/translations/en-us.yaml [hunk 1]
```diff
@@ -1444,7 +1444,11 @@ catalog:
           placeholder: 'default: 5'
     refreshInterval:
       label: Refresh Interval
-      placeholder: 'default: {hours}'
+      placeholder: 'default: {value}{unit}'
+      tooltip: Using any negative value disables Refresh
```
Comment (aruiz14) [inline shell/assets/translations/en-us.yaml]: This was actually rejected during earlier reviews of the original PR due to bad UX:
https://github.com/rancher/dashboard/pull/16963#issuecomment-4335377714

Comment (momesgin) [reply shell/assets/translations/en-us.yaml]: thanks for catching this!

**Review by oboc-sts — COMMENTED** [review]
The refresh interval changes are spot on @momesgin .

I didn't approve this just yet because I'm thinking maybe we can squeeze in a tiny improvement that could make this form look a bit more balanced and less fragmented (easier to scan).

So the question is: can we make the `name` field as long as the `URL` field below it to push the description a bit to the right (and resize it to make it fit) such that the "grid" stays on 2 major columns? It would be easier to follow through... 

Image

Wdyt?

**Comment by momesgin** [comment]
> Wdyt?

done:

image

**Review by oboc-sts — DISMISSED** [review]
Looks amazing! 🚀

File: shell/utils/time.ts [hunk 2]
```diff
@@ -143,3 +143,54 @@ export const dateTimeFormat = (value: string | undefined, store: Store<any>): st
...
+  }
+  if (seconds % 3600 === 0) {
+    return { value: seconds / 3600, unit: 3600 };
+  }
+  if (seconds % 60 === 0) {
+    return { value: seconds / 60, unit: 60 };
+  }
+
+  return { value: seconds, unit: 1 };
+}
```
Comment (marcelofukumoto) [inline shell/utils/time.ts]: **Suggestion:**
It would be interesting to have some guards for 0 and negative numbers.

File: shell/models/catalog.cattle.io.clusterrepo.js [hunk 3]
```diff
@@ -218,14 +219,36 @@ export default class ClusterRepo extends SteveModel {
...
+  }
+
+  get refreshIntervalDisplay() {
+    const val = this.spec?.refreshInterval;
+
+    if (val < 0) {
+      return this.t('generic.disabled');
+    }
+
+    return formatDuration(val || this.defaultRefreshInterval);
```
Comment (marcelofukumoto) [inline shell/models/catalog.cattle.io.clusterrepo.js]: **Question**
On this case, val as 0 will be treat as false case and use the default. If that is ok and expected, that is ok.
If it should use 0, than it needs to be fixed. But I guess 0 and nothing is same for this case.

File: shell/edit/catalog.cattle.io.clusterrepo.vue [hunk 4]
```diff
@@ -114,6 +133,27 @@ export default {
...
+        { label: this.t('unit.day', { count: 2 }), value: 86400 },
+      ];
+    },
+    refreshIntervalRules() {
+      return [
+        (val: string | number) => Number(val) < 0 ? this.t('validation.number.isPositive', { key: this.t('catalog.repo.refreshInterval.label') }) : undefined,
+      ];
+    },
+    refreshIntervalPlaceholder() {
+      const defaultHours = this.clusterRepoType === CLUSTER_REPO_TYPES.OCI_URL ? 24 : 1;
```
Comment (marcelofukumoto) [inline shell/edit/catalog.cattle.io.clusterrepo.vue]: **Question**
I believe since SUSE App Co repositories are OCI, it should be added here as 24?

File: shell/edit/catalog.cattle.io.clusterrepo.vue [hunk 5]
```diff
@@ -199,6 +259,9 @@ export default {
       this.ociMinWait = undefined;
       this.ociMaxWait = undefined;
       this.ociMaxRetries = undefined;
+      this.refreshEnabled = true;
+      this.refreshDisplayValue = null;
+      this.refreshUnit = 3600;
     },
     resetHelmValues() {
       delete this.value.spec['url'];
```
Comment (marcelofukumoto) [inline shell/edit/catalog.cattle.io.clusterrepo.vue:268]: **Suggestion**
The clear refresh is added to the gitRepo and Oci, but not the Helm one.
I tried to check what happens, and it always trigger 2 of them, so that is why always ok, if you have 3 and 2 of them have the reset than it will always apply. It doesnt cause a bug.
But would be interesting to create a function and always execute it as required on all the rests.

**Comment by marcelofukumoto** [comment]
I did some tests:

| Before | After |
| ------------- | ------------- |
| image  | image |

Added new field:
Details of UX:
- It turns disabled when checkbox is not checked
- Default to checked
- Whenever we change tabs it resets to checked
- The default written on all types are 1 hour, only OCI is written as 24hours, since SUSE App Co as well is an OCI, maybe it should be 24h
- Changed the Name width, adding a new property

Test are good, doesnt cover edge cases with 0, but all great tests.

File: shell/utils/time.ts [hunk 6]
```diff
@@ -143,3 +143,54 @@ export const dateTimeFormat = (value: string | undefined, store: Store<any>): st
 
   return day(value).format(format);
 };
+
+const DURATION_UNITS = [
+  { divisor: 86400, suffix: 'd' },
+  { divisor: 3600, suffix: 'h' },
+  { divisor: 60, suffix: 'm' },
+  { divisor: 1, suffix: 's' },
```
Comment (marcelofukumoto) [inline shell/utils/time.ts]: **Nit**
Maybe here we can use the values from the shell/utils/duration.js?
Having both of them make me wonder if the functions and things added here would work better there.
It would need to convert duration.js to TS though.

But would be interesting to create a const here anyway:
```
export const SECONDS_PER = {
  s: 1,
  m: 60,
  h: 3600,
  d: 86400,
} as const;
```

Since you always use the same values on the shell/edit/catalog.cattle.io.clusterrepo.vue file.

**Review by marcelofukumoto — COMMENTED** [review]
I did some tests, seems to be working well!
Added some suggestions for:
- Edge case of 0
- The way we handle the `time` and `duration` file. But exporting the numbers would be good.
- 24 as default for SUSE App Co as well since it is OCI
- And the refresh for Helm values

**Comment by richard-cox** [comment]
I removed myself from the reviewers list. When the absolute power trio of Alejandro, Marcelo and Ovi are happy feel free to merge

File: shell/utils/time.ts (hunk 2 above)
Comment (momesgin) [reply shell/utils/time.ts]: done

File: shell/edit/catalog.cattle.io.clusterrepo.vue (hunk 4 above)
Comment (momesgin) [reply shell/edit/catalog.cattle.io.clusterrepo.vue]: done

File: shell/utils/time.ts (hunk 6 above)
Comment (momesgin) [reply shell/utils/time.ts]: done

File: shell/models/catalog.cattle.io.clusterrepo.js (hunk 3 above)
Comment (momesgin) [reply shell/models/catalog.cattle.io.clusterrepo.js]: good find, zero is covered now.

File: shell/edit/catalog.cattle.io.clusterrepo.vue (hunk 5 above)
Comment (momesgin) [reply shell/edit/catalog.cattle.io.clusterrepo.vue:268]: done

File: shell/edit/catalog.cattle.io.clusterrepo.vue [hunk 7]
```diff
@@ -477,4 +569,19 @@ export default {
   height: max-content;
   overflow: hidden;
 }
+
+.refresh-interval-controls {
+  display: flex;
+  align-items: flex-end;
+  gap: 16px;
```
Comment (marcelofukumoto) [inline shell/edit/catalog.cattle.io.clusterrepo.vue]: **Nit**
I normally use `var(--gap-md)` for those cases.

File: shell/edit/catalog.cattle.io.clusterrepo.vue [hunk 8]
```diff
@@ -114,6 +133,27 @@ export default {
...
+      return [
+        { label: this.t('unit.sec'), value: SECONDS_PER.s },
+        { label: this.t('unit.min'), value: SECONDS_PER.m },
+        { label: this.t('unit.hour', { count: 2 }), value: SECONDS_PER.h },
+        { label: this.t('unit.day', { count: 2 }), value: SECONDS_PER.d },
+      ];
+    },
+    refreshIntervalRules() {
+      return [
+        (val: string | number) => Number(val) < 0 ? this.t('validation.number.isPositive', { key: this.t('catalog.repo.refreshInterval.label') }) : undefined,
```
Comment (marcelofukumoto) [inline shell/edit/catalog.cattle.io.clusterrepo.vue]: **Nit**
This one can use the formRulesGenerator, isPositive.
From @shell/utils/validators/formRules.

**Review by marcelofukumoto — DISMISSED** [review]
I tested again, saved and the changes were good.
Just added 2 Nits which is most reusing functions. But not required.

LGTM

File: shell/edit/catalog.cattle.io.clusterrepo.vue (hunk 7 above)
Comment (momesgin) [reply shell/edit/catalog.cattle.io.clusterrepo.vue]: good point, done

File: shell/edit/catalog.cattle.io.clusterrepo.vue (hunk 8 above)
Comment (momesgin) [reply shell/edit/catalog.cattle.io.clusterrepo.vue]: thanks for reminding me of this, done

**Review by marcelofukumoto — DISMISSED** [review]
Thank you for the Nit changes.
LGTM

**Review by aruiz14 — DISMISSED** [review]
LGTM, thanks @momesgin 

Regarding the zero case, in my original PR I attempted to prevent introducing any negative or zero value, given they are interpreted differently when stored:
 - Negative -> disabled, which has a explicit checkbox now
 - Zero -> omitted when serialized -> replaced by blank

It was hard for to implement given my limited frontend experience, you'd probably find better ways to do it, or simply leave it as is if you think that's an acceptable UX.

**Comment by momesgin** [comment]
> LGTM, thanks @momesgin
> 
> Regarding the zero case, in my original PR I attempted to prevent introducing any negative or zero value, given they are interpreted differently when stored:
> 
> * Negative -> disabled, which has a explicit checkbox now
> * Zero -> omitted when serialized -> replaced by blank
> 
> It was hard for to implement given my limited frontend experience, you'd probably find better ways to do it, or simply leave it as is if you think that's an acceptable UX.

Thanks for the review as well 🙏 I think the zero case can be addressed separately, also the backend should cover the zero case as well (if it doesn't already) since obviously the endpoint can be called outside of UI too.

**Review by marcelofukumoto — DISMISSED** [review]
Checked again after the master merge.
LGTM

ps.: I checked the 0 and null case. (null being no key, not value null)

If you add 0 the zero is added on the data.
image

If you edit again it show as empty even with the ZERO on the data.
image

if you save without changing it will keep ZERO, even if you send the empty field.
If you change from hour to seconds for example it will send NULL again.

It is a little inconsistent but at the same time ZERO and NULL should behave the same on the backend and use the default as I could understand.

Here it will check if more than 0.
https://github.com/rancher/rancher/blob/599a9957e5726b0627373013ad56cfbe4bc60f4e/pkg/controllers/dashboard/helm/repo.go#L125

Here it defines as int not *int and with omitempty.
https://github.com/rancher/rancher/blob/599a9957e5726b0627373013ad56cfbe4bc60f4e/pkg/apis/catalog.cattle.io/v1/types.go#L79

Just did some simple analysis. As a Nit we can clear it up and never send 0 to be consistent on the API request. But it seems they behave the same.

**Comment by momesgin** [comment]
@marcelofukumoto thanks a lot for analyzing the backend code as well as the UI's 🙏

**Review by aruiz14 — APPROVED** [review]
